### ISSUE TITLE
Updatable Activity Classifier in Turi not giving good results

### ISSUE THREAD

**Issue by christinalau** [apple/turicreate#2593]
I would like to create an Updatable Activity Classifier that will allow racquet sports users to create their own personalized CoreML model to detect 5 swing motion (forehand, backhand, smash, forehand volley and backhand volley) for different racquet sports including tennis, padel, squash, badminton, pickelball and racquetball using their Apple Watch. I would like to add this new feature to [TennisKeeper](https://apps.apple.com/app/tenniskeeper/id1097388824), a racquet sports tracker, so that users can improve the accuracy of their shot tracking.

My overall approach is as follows:

1. Use Turi create to create a StrokeClassifier.mlmodel using a DataSet1 which contains 1 of each 5 swing motion type.
2. Use coremltools to make it updatable: StrokesUpdatable.mlmodel.
3. Import both into XCode 11.2.1 and run a number of tests.

Currently, I am not getting expected results with my tests. Here is a brief summary:

**Test 1:** Run predict on StrokesClassifier and StrokesUpdatable with the same DataSet1. Gets 100% recognition _(Expected)_.

**Test 2:** Use a new DataSet2 which has all 5 stroke types, they are similar to DataSet1, but not identical swing motion.
Run predict with DataSet2 on StrokesClassifier and StrokesUpdatable. I got the same results, both with less than 100% prediction _(Expected)_.

**Test 3**: Use MLUpdateTask to update StrokesUpdatable with DataSet2.
Run predict again on StrokesUpdatable with DataSet2. This time I expect the results to be better than Test 2, but I am not getting that. It is as if the MLUpdateTask did nothing to update the model. _(Unexpected)_.

**Test 4:** Use an almost empty DataSet with only 1 Unknown stroke type (i.e. No forehand, backhand, smash, forehand volley, backhand volley) and create a new StrokesUpdatable. This time when I use MLUpdateTask to update the model with DataSet2. The
`updateContext.model.write` call failed with this error: 

      The operation couldn’t be completed. (Foundation._GenericObjCError error 0.)

It seems if the original model does not already have at least 1 value for each Class label (i.e. forehand, backhand etc) the write will fail. _(Unexpected)._

Can you help me understand what is going on here? How can I debug the updated model?

Ideally I will like to initialize a model with none of my own strokes data and let users personalize it completely, but Test 4 seems to suggest that it is not possible. 

Thanks for any help.

**Comment by anilkatti** [comment]
Hi @christinalau ! I will help you debug this. There are a lot of subtleties here so, please bear with me. Let's start with your Test 3. Could you please tell me:

1. Accuracy with DataSet2 before starting update process.
2. Number of epochs you ran this training for
3. Learning rate you used in this case
4. Accuracy with DataSet2 after the update was complete.
5. Value of loss at the end of each epoch during the training.

**Comment by christinalau** [comment]
Hi @anilkatti , can you tell me how to get these data from XCode iOS? I don't seem to find such APIs. I found this page:  [Model Performance](https://developer.apple.com/documentation/createml/improving_your_model_s_accuracy) but it is for Create ML only, not Core ML. Hence I don't even have the `evaluation` or `trainingMetrics` API in my XCode project. I may be missing something obvious here.

If you want me to do that in Turi, can you tell me how I can consume this updated model in Turi?

![personalizedmodel](https://user-images.githubusercontent.com/5726085/68671593-ab0a7080-051d-11ea-9ff4-b70c238d9abb.png)

Sorry to ask so many questions but I can't seem to google my way out of this :-(

**Comment by anilkatti** [comment]
@christinalau no worries.. let me clarify:

> Accuracy with DataSet2 before starting update process.

I meant prediction accuracy. What percent of DataSet2 did the model get right before update?

> Number of epochs you ran this training for

This should be visible in the Xcode UI when you double-click on the model. If you changed it at runtime by setting a value against MLParameterKey.epochs in configuration.parameters, let me know what that is. Lastly, you could set a callback to be called at end of every epoch. So, you could count number of epochs that way as well.

> Learning rate you used in this case

Same here. You should see it in the Xcode UI for the model.

> Accuracy with DataSet2 after the update was complete.

Prediction accuracy after the model was updated.

> Value of loss at the end of each epoch during the training.

You can track this in update task callbacks. context.metrics[MLParameterKey.learningRate].

**Comment by anilkatti** [comment]
@christinalau Alternatively, attach a sample project with the model and DataSet2 (if it is not sensitive) and I can debug locally.

**Comment by christinalau** [comment]
@anilkatti Thank you for the clarification. Here are the answers:

1. Accuracy before update process: 86%
2. Number of epochs: 10
3. Learning Rate: 0.01  (Note both are from original model, I did not make change to them).
4. Accuracy after update process: 86%. In fact it is exactly the same prediction as before update.
5. `updateContext.metrics[MLMetricKey.lossValue]` return 1.595125

Just for fun, I ran update 3 times with the same DataSet2, the prediction came back identical, but the lossValue has changed as follows (1.577465, 1.561244, 1.546327). Not sure if this means anything or not. Thanks...

**Comment by christinalau** [comment]
I have attached a zip file with model, original data (Under Training Data Folder), and DataSet2 (under TestFiles). 
[DebugUpdate.zip](https://github.com/apple/turicreate/files/3843720/DebugUpdate.zip)

**Comment by anilkatti** [comment]
Sorry about the delay.

> Just for fun, I ran update 3 times with the same DataSet2, the prediction came back identical, but the lossValue has changed as follows (1.577465, 1.561244, 1.546327). Not sure if this means anything or not. Thanks...

You tried a really good experiment. Clearly, loss is going down but, not as fast as you'd want to make a difference in accuracy.

This means we need to tune the hyper-parameter for this model. Could you try increasing learning rate? You could try different values - like 0.02, 0.03, 0.05, 0.1, 0.15, etc. Here's how you'd do that:

```
let configuration = MLModelConfiguration()

// requests training loop run for 10 iterations and use 0.02 learning rate
configuration.parameters = [MLParameterKey.epochs : 10, MLParameterKey.learningRate : 0.02]

let progressHandlers = MLUpdateProgressHandlers(forEvents: .epochEnd,
                                                progressHandler: progressHandler,
                                                completionHandler: completionHandler)

let updateTask = try MLUpdateTask(forModelAt: updatableModelURL,
                                  trainingData: trainingData,
                                  configuration: configuration,
                                  progressHandlers: progressHandlers)

// start the update process
updateTask.resume()
```

`progressHandler` would look something like:

```
func progressHandler(_ context: MLUpdateContext) {
    // print context.metrics[MLMetricKey.loss]
}
```

See if loss goes down for each epoch.

Here's more details on the API usage: https://github.com/apple/coremltools/blob/master/examples/updatable_models/OnDeviceTraining_API_Usage.md

**Comment by christinalau** [comment]
@anilkatti  Here is what I found. As Learning Rate goes up, loss goes down, but accuracy also goes down. Not quite expected?

```
Learning Rate = 0.02   Accurray = 86%

Value Loss:
1.612373
1.607763
1.621093
1.606253
1.597315
1.621052
1.601825
1.594099
1.602741
1.588197


Learning Rate = 0.03	Accurray = 86%

Value Loss:
1.612373
1.605737
1.617471
1.600861
1.59012
1.612564
1.591749
1.583164
1.590842
1.574894



Learning Rate = 0.05	Accurray = 82%

Value Loss:

1.612373
1.60171
1.610376
1.590423
1.576365
1.596537
1.572957
1.563064
1.56924
1.551025


Learning Rate = 0.1		Accurracy = 82%

Value Loss:

1.612373
1.591794
1.593475
1.566246
1.545392
1.561451
1.532951
1.521634
1.525887
1.504307


Learning Rate = 0.15	Accurracy = 50%

Value Loss:

1.612373
1.582091
1.577719
1.544583
1.518737
1.53242
1.501116
1.49013
1.494055
1.471145



Learning Rate = 0.50	Accurracy = 50%

Value Loss:

1.612373
1.520023
1.494251
1.444498
1.410475
1.427172
1.396448
1.397576
1.405532
1.38536


Learning Rate = 1.0    Accurracy = 50%

Value Loss:

1.612373
1.448231
1.43071
1.385805
1.35816
1.386585
1.357867
1.366183
1.374959
1.358988
```

**Comment by anilkatti** [comment]
@christinalau could you attach a sample code for this? Just to clarify, you could not mark dense as updatable, correct?

**Comment by christinalau** [comment]
@anilkatti  That is correct. Per our other discussion, I did: 
`builder.make_updatable(['dense_layer1'])`
to make the model updatable.

I have attached the 2 classes for update and prediction.
[DebugCode.zip](https://github.com/apple/turicreate/files/3852903/DebugCode.zip)

**Comment by anilkatti** [comment]
@christinalau thanks for providing that. I spent quite some time today trying to make sense of this. Here's what I think is going on:

The last inner product layer in your model has a small number of parameters. When we increase the learning rate, the network is forgetting some things it previously learnt. It is called "Catastrophic Forgetting in Neural Networks" https://arxiv.org/abs/1708.02072

It is unfortunate that marking inner product layer before batchnorm as updatable fails to compile. I am trying out a few things to see if that helps your use case.
